samples/wifi/provisioning/implementation_details.rst

```diff
@@ -0,0 +1,268 @@
+If invalid argument is provided, the status shall be STATUS_INVALID_ARGUMENT.
+If internal error occurs, the status shall be STATUS_INTERNAL_ERROR.
+
+If the operation is successful, the device shall send state update to configurator.
+The message shall be in type of Result defined in result.proto and encoded.
+The updated state of Wi-Fi should be in state field. If the state is CONNECTION_FAILED, the device may fill reason field, whose value depends on the failure reason.
+If the state is CONNECTED, the device shall store the credentials in NVM.
+
+FORGET_CONFIG procedure
+-----------------------
+
+When request with Op Code of FORGET_CONFIG is received, the device shall remove stored credentials from NVM, and depending on the result, a response message is sent to configurator.
+
+If the operation is successful, the device shall send response message with status of STATUS_SUCCESS.
+And the device shall disconnect from access point if its state is connected.
+If internal error occurs, the status shall be STATUS_INTERNAL_ERROR.
+
+You may find the details of definition in .proto files under path of ./common/proto.
+
+Message delivery
+================
+
+To exchange messaged mentioned above, some sort of transport method is required.
+Besides, it should also provide a certain level of security.
+In our implementation, Bluetooth Low Energy (BLE) is chosen.
+It is widely deployed and available on most smartphones, which play the role of configurator.
+Meanwhile, we leverage its capability of information confidentiality and integrity.
+BLE itself has pairing function, and can provide different level of security.
+GATT allows us to assign different security level to different characteristic under the same service, making fine tune of access control possible.
+
+Service declaration
+-------------------
+
+The Wi-Fi Provisioning Service is instantiated as a Primary Service.
+The service UUID shall be set to the UUID value as defined in tables below.
+
+========================== ====================================
+Service Name               UUID
+Wi-Fi Provisioning Service 14387800-130c-49e7-b877-2881c89cb258
+========================== ====================================
+
+Service characteristics
+-----------------------
+
+The UUID value of characteristics are defined in table below.
+
+========================== ====================================
+Characteristic Name        UUID
+Information                14387801-130c-49e7-b877-2881c89cb258
+Operation Control Point    14387802-130c-49e7-b877-2881c89cb258
+Data Out                   14387803-130c-49e7-b877-2881c89cb258
+========================== ====================================
+
+The characteristic requirements of the Wi-Fi Provisioning Service are shown in table below.
+
++-----------------+-------------+-------------+-------------+-------------+
+| Characteristic  | Requirement | Mandatory   | Optional    | Security    |
```

Comment:
What does M mean? Can we use yes/no for requirements?

Security and Required _in table items_ do not need to be capitalized.

Reply:
If there are no optional properties, just leave out the column.

I think we also don't need the Requirements table, really, because we said in the intro text that these are requirements.

Reply:
M stands for mandatory. This is how the bluetooth specs express. But you are right that the abbreviation is used without explanation. For optional properties column, I prefer to keep it, as tables are the same in every bluetooth service spec.